### PR TITLE
Fix unreliable logic to update site URLs registered on authentication service

### PR DESCRIPTION
**Pull request by benbowler**
## Summary

Addresses issue #2383

## Relevant technical choices

Although PHP allows comparison of two arrays using `==` I chose to compare by looping through the get_site_fields and matching each to the fetched values so that we can match for type as well as value while still ignoring the order of the items in the array.

```
foreach ( $get_site_fields as $key => $site_field ) {
	if ( ! array_key_exists( $key, $fetch_site_fields ) || $fetch_site_fields[ $key ] !== $site_field ) {
		return false;
	}
}
```

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Hi @benbowler. Is this ready for review? I don't see the following requirement from IB to be implemented. Perhaps you forgot to push your changes?

```
Update the Authentication::handle_sync_site_fields method to call $this->google_proxy->sync_site_fields( ... ) only if $this->google_proxy->are_site_fields_synced( ... ) returns FALSE;
```

**Comment by benbowler** [comment]
You're right @eugene-manuilov, I've updated that function as per the IB.

File: includes/Core/Authentication/Authentication.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Authentication.php:1155]: @benbowler let's explicitly define that we need to sync site fields when the `are_site_fields_synced` method returns `FALSE` because it can also return an instance of the `WP_Error` class.

```suggestion
		if ( $this->google_proxy->are_site_fields_synced( $this->credentials ) === false ) {
```

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:102]:
```suggestion
	 * @since n.e.x.t
```

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:144]:
```suggestion
	 * @since n.e.x.t
```

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (eugene-manuilov) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:47]: @benbowler let's also add a case when proxy returns an error to make sure that `fetch_site_fields` returns an appropriate WP_Error instance.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (eugene-manuilov) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:146]: @benbowler no need to assert functionality from the `fetch_site_fields` method. It is already checked in its own test.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (eugene-manuilov) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:170]: @benbowler let's use `$this->assertWPErrorWithMessage`, defined in the `tests/phpunit/includes/TestCase.php` file, to better evaluate the error object.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler. It's a good start. Added some minor comments to address before I can approve it.

**Comment by benbowler** [comment]
Thanks @eugene-manuilov, all updated.

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:108]: @benbowler let's return an error if we don't have credentials.

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:124]: We need to check the response object itself, not its body:

```suggestion
		if ( is_wp_error( $response ) ) {
			return $response;
		}
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler! Just two more minor things to address and I'll approve it.

File: includes/Core/Authentication/Google_Proxy.php
Comment (benbowler) [reply includes/Core/Authentication/Google_Proxy.php:124]: This does not work because the $reponse is a standard array, the WP_Error or other results are returned in the body of the request. In the initial PR I fixed this by testing `is_wp_error($raw_body)`, we can do this or test the `$reponse['body']` let me know which you prefer.

```
var_dump($response);
```

```
array(5) {
  ["headers"]=>
  array(0) {
  }
  ["body"]=>
  object(WP_Error)#3664 (2) {
    ["errors"]=>
    array(1) {
      ["test_message"]=>
      array(1) {
        [0]=>
        string(12) "test_message"
      }
    }
    ["error_data"]=>
    array(0) {
    }
  }
  ["response"]=>
  array(2) {
    ["code"]=>
    int(200)
    ["message"]=>
    string(2) "OK"
  }
  ["cookies"]=>
  array(0) {
  }
  ["http_response"]=>
  NULL
}
```

Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php:124]: @benbowler see the documentation for the [`wp_remote_post`](https://developer.wordpress.org/reference/functions/wp_remote_post/) function:

```
Return #
(array|WP_Error) The response or WP_Error on failure.
```

**Comment by benbowler** [comment]
Hey @eugene-manuilov, I've updated the WP_Error for oauth_credentials_not_exist. For the other request, it's not possible to check the raw `$response` array `is_wp_error`, I've left a comment above and reverted to my original implementation of checking `$raw_body` `is_wp_error`.

**Comment by benbowler** [comment]
Updated.

**Comment by benbowler** [comment]
@eugene-manuilov

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:108]: @benbowler, please, add `use WP_Error;` to the top of the class to make it work here.

File: includes/Core/Authentication/Google_Proxy.php
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:126]: @benbowler let's also make sure that the response status code is 200:

```php
if ( wp_remote_retrieve_response_code( $response ) != 200 ) {
    return new WP_Error( ... );
}

$raw_body = wp_remote_retrieve_body( $response );
```

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (eugene-manuilov) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:74]: @benbowler let's better use the real URL for it. You can create another helper function to test errors. Something like this:

```php
private function mock_http_failure( $request_url, $response_error ) {
    add_filter(
        'pre_http_request',
        function( $response, $parsed_args, $url ) use ( $request_url, $response_error ) {
            if ( $url === $request_url ) {
                return new WP_Error( $response_error );
            } else {
                return $response;
            }
        },
        10,
        3
    );
}
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler. Found a few more things to improve. Could you please check it?

**Review by eugene-manuilov — DISMISSED** [review]
LGTM! Thanks, @benbowler!

File: includes/Core/Authentication/Google_Proxy.php
Comment (felixarntz) [inline includes/Core/Authentication/Google_Proxy.php:129]: This is somewhat unprecise, as the `WP_Error` returned would not provide information on the real error here - which may be okay as a last resort, but we should try to use more specific errors. For example, below you run the check whether the response contains an `error` key, which is possible, and in that case the response status code would not be 200. So we'd want to in such a case return the code in `error` rather than a less precise error that the response didn't come back with a 200.

I think we should just remove this clause completely. If the entire request fails, that would be caught by the `WP_Error` from `wp_remote_post`, and if the request comes back with an error response, we should be okay relying on the `error` key in the response.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (felixarntz) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:354]: Why was this changed? I don't see any usage where `$response_data` is not an array, and I think it's better to keep the signature consistent, especially if there's no reason to change it.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (felixarntz) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:373]: This shouldn't be in parentheses, and the parameter docs should be aligned.
```suggestion
	 * @param string   $request_url    Request URL to modify response for.
	 * @param WP_Error $response_error Response WP_Error to return.
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
Great! Just some minor feedback on things that could be improved.

Normally I wouldn't block merging on such changes, but since this is still awaiting for the authentication service side to be launched, I think it's worth making these minor enhancements.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (benbowler) [reply tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:354]: In a previous commit, I updated this function to allow tests to pass a WP_Error via $response_data before splitting it out later into `mock_http_failure`.

I've added this type back in. 👌

File: includes/Core/Authentication/Google_Proxy.php
Comment (benbowler) [reply includes/Core/Authentication/Google_Proxy.php:129]: Great, updated and pushed in the most recent push.

File: tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php
Comment (felixarntz) [inline tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:350]: The check here is also no longer necessary, since it'll always be an array. We can unconditionally `json_encode` it like before.

**Review by felixarntz — APPROVED** [review]
Awesome! Logic on our end is now pushed to the staging version of the authentication service, and I've updated the QA Brief with the related information for testing.

**Pull request merged**
